File: 389. Find the Difference.cpp
class Solution {
public:
    // hashtable: O(N) time, O(1) space, 44.29%, ok
    char findTheDifference(string s, string t) {
        vector<int> letter(26, 0);
        for(int i = 0; i < s.size(); i++){
            letter[s[i] - 'a']++;
            letter[t[i] - 'a']--;
        }
        letter[t[t.size() - 1] - 'a']--;
        for(int i = 0; i < 26; i++)
            if(letter[i] < 0)
                return 'a' + i;
        return 'a';
    }

    // bit manipulation: O(N) time, O(1) space, 44.29%, ok
    char findTheDifference(string s, string t) {
        char ret = 0;
        for(auto c: s) ret ^= c;
        for(auto c: t) ret ^= c;
        return ret;
    }
};

File: 389. Find the Difference.py
class Solution(object):
    
    # Hashtable: O(m) time, O(m) space, 45.48%
    def findTheDifference(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        count = {}
        for c in s:
            count[c] = count.get(c, 0) + 1
        for c in t:
            count[c] = count.get(c, 0) - 1
            if count[c] < 0:
                return c
      
    # Bit Manipulation: O(m) time, O(1) space, 57.66%
    def findTheDifference(self, s, t):
        # return chr(reduce(lambda x, y: x ^ y, map(lambda x: ord(x), s + t)))
        return chr(reduce(operator.xor, map(ord, s + t)))  # 88.31%